alica_engine/src/engine/RunnableObject.cpp

```diff
@@ -50,7 +53,46 @@ void RunnableObject::start(RunningPlan* rp)
     }
     ++_signalState;
     _signalContext.store(rp);
-    _engine->editScheduler().schedule(std::bind(&RunnableObject::doInit, this));
+    if(_requiresParameters) {
+        assert(rp->getParent());
```

Comment:
https://github.com/rapyuta-robotics/alica/pull/187#discussion_r759868156

Doing this in RunningPlan would be cleaner and you can avoid circular dependency. RunningPlan is now calling basicPlan->start which then again calls methods of RunningPlan. It can be easily avoided in this case.

Reply:
RunningPlan doesn't call basicPlan->start directly.  It calls a function of PlanPool or BehaviorPool (depending on type of child) which then  looks up the appropriate RunnableObject type based on the configuration in the running plan and then calls start().   The difficult is that this means we don't actually have access to the appropriate RunnableObject when we would generate this from the RunningPlan.  And although we could pass down the blackboard of the parent, we would still need to access the parent RunningPlan here to find the appropriate plan attachment.  That's why I didn't think it would be worth doing that refactoring

Reply:
We can pass 2 arguments, blackboard of parent (that's needed regardless of whether child shares the parents blackboard or not), optional attachment function of parent if the plan requires parameter. Rather than looking up the tree, a RunningPlan can even pass this information while creating a child RunningPlan.  I am trying to avoid circular dependency here to make our life easier and help us in future cleanups.

Reply:
https://www.wrike.com/open.htm?id=807830766
https://www.wrike.com/open.htm?id=807830140
I think it is far easier to refactor first and then ensure that RunningPlan does not need to reference its parent, due to the different lifecycles of RunningPlan and RunnableObject